### PR TITLE
fix: add explicit SMul instance in Finsupp

### PR DESCRIPTION
**Pull request by kbuzzard**
This makes the group cohomology files faster. Still plenty to be done but this is a start.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by kbuzzard** [comment]
On my machine this reduces the time taken (as reported by reported by `trace.profiler` ) to compile `GroupCohomology.Resolution.ofMulActionBasisAux` from 45 seconds to 15.

**Comment by kbuzzard** [comment]
!bench

**Comment by kbuzzard** [comment]
Clarification: the commits since the benchmark are just me dividing various maxheartbeat values in group cohomology files by 2.

File: Mathlib/Data/Finsupp/Basic.lean [hunk 1]
```diff
@@ -1471,8 +1471,10 @@ end
...
 section
 
-instance smulZeroClass [Zero M] [SMulZeroClass R M] : SMulZeroClass R (α →₀ M) where
+instance smul [Zero M] [SMulZeroClass R M] : SMul R (α →₀ M) where
   smul a v := v.mapRange ((· • ·) a) (smul_zero _)
+
+instance smulZeroClass [Zero M] [SMulZeroClass R M] : SMulZeroClass R (α →₀ M) where
   smul_zero a := by
     ext
     apply smul_zero
```
Comment (eric-wieser) [inline Mathlib/Data/Finsupp/Basic.lean:1480]: I would be surprised (and a bit sad) if this helps, but I guess I was surprised by the Kaehler PR working too...

Did you ever test this separately from the other two changes, to get a feel for which is most significant?

Comment (kbuzzard) [reply Mathlib/Data/Finsupp/Basic.lean:1480]: I'm in the process of doing some extensive testing now. Will report back.

**Review by kbuzzard — COMMENTED** [review]
Let's see if #6189 gives the same speedups with fewer changes to `Data.Finsupp.Basic`.

**Comment by kbuzzard** [comment]
Thanks Eric, you're right: it is only the `module` change which has any noticeable effect; conversely the `module` change still enables all the same heartbeat-lowering in the group cohomology files (at least locally). Closing in favour of #6189 .

**Pull request closed without merging**